src/core/reducers/uiState.js

```diff
@@ -0,0 +1,129 @@
+/* @flow */
+import * as React from 'react';
+import invariant from 'invariant';
+import { compose } from 'redux';
+import { connect } from 'react-redux';
+
+import { normalizeFileNameId } from 'core/utils';
+
+export const SET_UI_STATE = 'SET_UI_STATE';
+
+type UIStateType = {
+  [id: string]: Object,
+};
+
+const initialState = {};
+
+type SetUIStateParams = {|
+  id: string,
+  change: Object,
+|};
+
+type SetUIStateAction = {|
+  payload: SetUIStateParams,
+  type: typeof SET_UI_STATE,
+|};
+
+export const setUIState = ({
+  change,
+  id,
+}: SetUIStateParams): SetUIStateAction => {
+  invariant(change, 'change cannot be undefined');
+  invariant(id, 'id cannot be undefined');
+
+  return { type: SET_UI_STATE, payload: { change, id } };
+};
+
+export const generateId = ({
+  fileName,
+  id,
+}: {|
+  fileName: string,
+  id: string,
+|}) => {
+  invariant(fileName, 'fileName is required');
+  invariant(typeof id === 'string', 'id must be a string');
+
+  return `${normalizeFileNameId(fileName)}-${id}`;
+};
+
+export const withUIState = ({
+  fileName,
+  extractId,
+  defaultState = {},
+}: {|
+  fileName: string,
+  extractId: (props: Object) => string,
+  defaultState?: Object,
+|}): ((React.ComponentType<any>) => React.ComponentType<any>) => {
+  invariant(fileName, 'fileName is required');
+  invariant(extractId, 'extractId is required');
+  invariant(defaultState, 'defaultState is required');
+
+  const mapStateToProps = (
+    state: {| uiState: UIStateType |},
+    props: Object,
+  ) => {
+    const uiStateID =
+      props.uiStateID || generateId({ fileName, id: extractId(props) });
+    const uiState = state.uiState[uiStateID] || defaultState;
```

Comment:
Do we really need a `defaultState`? I wish we had a `initialState` instead, similar to:

```
constructor(props)

Reply:
Let me try to explain the problem from the perspective of `EditableCollectionAddon`.

It was doing this to initialize state:

```js
constructor(props) {
  super(props);
  this.state = {
    editingNote: false,
  };
}
```

I introduced `withUIState({ defaultState: ... })` to mimic this but because Redux state is persistent across all component instances (because `extractId()` only looks at props), I needed this to reset the state for *every* new instance:

```js
componentWillMount() {
  this.props.setUIState({ editingNote: false });
}
```

This allowed it to function like before.

I can think of some different solutions here. Switching to `componentDidMount()` is an obvious solution but setting an initial state really isn't an async operation so let me think of some other approaches.

Reply:
Yes, I figured out that my code snippet above fixed only one part of the problem, but not exactly what I was highlighting. We need something to initialize the state in `uiState` before the first `render()`.

Reply:
mmh, that might work actually since the props are initialized with `defaultState`. We always have valid data injected by the HOC. By spreading the `stateProps`, the first call to `setUIState()` should add the `defaultState` overrode by `change` into redux. That should reproduce the behavior of `this.state` into the constructor as well as what `setState()` does.